identity/src/public_key.rs

```diff
@@ -13,63 +13,122 @@
+            Secp256k1(pk) => pk.verify(msg, sig.to_bytes().as_slice())
+        }
+    }
 
-impl Debug for PublicKey {
-    fn fmt(&self, f: &mut ::core::fmt::Formatter<'_>) -> ::core::fmt::Result {
-        write!(f, "{:?}", self.0)
+    pub fn encode(&self) -> Vec<u8> {
+        use PublicKey::*;
+        let mut result: Vec<u8> = Vec::new();
+
+        result.push(self.get_prefix());
+        match self {
+            Ed25519(pk) => result.extend(pk.encode().to_vec()),
+            #[cfg(not(target_arch = "wasm32"))]
+            Rsa(pk) => result.extend(pk.encode_pkcs1().to_vec()),
+            Secp256k1(pk) => result.extend(pk.encode().to_vec()),
+        };
+
+        result
     }
-}
 
-impl PublicKey {
-    pub fn verify_strict(
-        &self,
-        message: &[u8],
-        signature: &Signature,
-    ) -> Result<(), SignatureError> {
-        self.0.verify_strict(message, &signature.0)
+    pub fn decode(bytes: Vec<u8>) -> Result<PublicKey, DecodingError> {
+        match bytes[0] {
+            0 => Ok(PublicKey::Ed25519(ed25519::PublicKey::decode(&bytes[1..])?)),
+            1 => Ok(PublicKey::Rsa(rsa::PublicKey::decode_pkcs1(&bytes[1..])?)),
+            2 => Ok(PublicKey::Secp256k1(secp256k1::PublicKey::decode(&bytes[1..])?)),
+            _ => Err(DecodingError::new("invalid type byte".to_string())),
+        }
     }
 
-    pub fn from_base58(str: &str) -> Result<PublicKey, PKError> {
-        let bytes = bs58::decode(str).into_vec().map_err(FromBase58Error)?;
-        Self::from_bytes(&bytes)
+    fn get_prefix(&self) -> u8 {
+        use PublicKey::*;
+        match self {
+            Ed25519(_) => 0,
+            #[cfg(not(target_arch = "wasm32"))]
+            Rsa(_) => 1,
+            Secp256k1(_) => 2
+        }
     }
 
-    pub fn from_bytes(bytes: &[u8]) -> Result<PublicKey, PKError> {
-        let pk = ed25519_dalek::PublicKey::from_bytes(bytes).map_err(FromBytesError)?;
-        Ok(PublicKey(pk))
+    pub fn from_base58(str: &str) -> Result<PublicKey, DecodingError> {
+        let bytes = bs58::decode(str).into_vec().map_err(DecodingError::new)?;
+        Self::decode(bytes)
     }
 
-    pub fn to_bytes(&self) -> [u8; ed25519_dalek::PUBLIC_KEY_LENGTH] {
-        self.0.to_bytes()
+    pub fn to_bytes(&self) -> Vec<u8> {
+        use PublicKey::*;
+
+        match self {
+            Ed25519(pk) => pk.encode().to_vec(),
+            Rsa(pk) => pk.encode_pkcs1().to_vec(),
+            Secp256k1(pk) => pk.encode().to_vec(),
+        }
     }
+}
+
+impl From<libp2p_core::identity::PublicKey> for PublicKey {
+    fn from(key: libp2p_core::identity::PublicKey) -> Self {
+        use libp2p_core::identity::PublicKey::*;
 
-    pub fn from_libp2p(pk: &identity::ed25519::PublicKey) -> Result<Self, PKError> {
-            Self::from_bytes(&pk.encode())
+        match key {
+            Ed25519(key) => PublicKey::Ed25519(ed25519::PublicKey::decode(key.encode().to_vec().as_slice()).unwrap()),
```

Comment:
No reason for cloning, let's drop `to_vec`
```suggestion
            Ed25519(key) => PublicKey::Ed25519(ed25519::PublicKey::decode(&key.encode()).unwrap()),
```